### DISCUSSION TITLE
MESのジョブディスパッチ機能対応検討

### DISCUSSION THREAD

**Discussion by ta-ichihashi** [Beckhoff-JP/PLC_JobManagementFramework#3] [Ideas]
4Mのうちの、Materialリソースオブジェクトを取り扱う目的で、加工ジョブ（process job）を取り扱えるようにしたい。一方で機械の各モジュールに対する動作指示をMachine（搬送や加工）リソースとしてジョブ管理しなければならない。MaterialジョブからはいつでもMachineジョブが使えるわけではなく、リソースの空き状況により実行可否を制御しなければならない。
よって、Materialジョブ側の各実行モジュールのFB_Executor.start()を実行するタイミングを制御して、Machineリソースの空きに従ってジョブ実行を制御たい。

オブジェクト構造としては以下が考えられる。

Materialジョブの場合

``` 
PararelJobContainer    材料（ワーク）の処理を取り扱うためのコンテナ
   |
   +- BatchJobContainer　材料単位の順次実行ジョブコンテナ
   |
   +- BatchJobContainer
                 :
```

Machineジョブの場合

```
PararelJobContainer    機械（搬送装置や処理モジュール）を取り扱うためのコンテナ
   |
   +- QueueJobContainer　排他が必要な機械モジュールごとのジョブコンテナ
   |
   +- QueueJobContainer
                     :
```

以上をもとに、次の機能拡張が必要。

* PararelJobContainer, QueueJobContainerについて、包含している_executors全て完了したら自動的にquit()、finish()するのではなく、process状態を保持するモードを作る。FB_Executor.quit()を追加して、これにより終了コールできるようにする。（それまではジョブがなくなったあともadd_future()でジョブを追加できる）
* 並行して動作するMaterialジョブがMachineジョブにキューインする優先順を決定するための根拠となるデータの提供。
  * execute() の実行実績、ジッタの計測機能とクエリ機能など
* 代理実行機能
   * Materialジョブに機械を制御するタスクを登録しておき、実行時はMachineジョブにより代理実行する機構が必要。
   * Materialジョブには、複数のMachineコンポーネントが持つタスクが組み合わされて構成されている。
   * 代理実行機構は、実行時にこれらのタスクを分解し、機械モジュール毎に用意されたQueueJobContainerへキューインする。

**Comment by makitagawa243** [comment]
仮想ですがインスタンスレベルの例にするとこのようなユースケース想定でしょうか。

* システム構成
![image](https://github.com/Beckhoff-JP/PLC_JobManagementFramework/assets/11852566/00f42c0b-39e9-4e6f-964a-f773528f644c)
    * 品種によって洗浄・切削・圧入を自由に組み合わせて加工して払い出すシステム
    * ユニット間はロボット1式で搬送する

品種Aは洗浄→切削→洗浄→圧入
品種Bは洗浄→圧入→切削→洗浄

+ 上位からの加工ジョブ指示

    ````json
  {
    "product": "品種A",
    "quantity":10,
    "recipeId":1
  },
  {
    "product": "品種B",
    "quantity":20,
    "recipeId":2
  }
    ````

このジョブが来た場合、各ユニットのタクトが全て同一と仮定すると以下のような実行順になる。

| ジョブ指示 | 品種A開始 | 品種B開始 |     |     |     |     |
| ----- | ----- | ----- | --- | --- | --- | --- |
| 洗浄    |       | 品種A   | 品種B | 品種A |     | 品種B |
| 切削    |       |       | 品種A |     | 品種B |     |
| 圧入    |       |       |     | 品種B | 品種A |

* 品種Aのジョブを実行中に品種Bの生産指示が来た場合、洗浄ユニットは既に品種Aで利用中のため終了を待機する
* 品種Bの洗浄1回目完了後は、圧入ユニットは空いているため即実行が可能
* タクトが延びた場合は、ユニットの占有状態によってお互いのジョブが待ち合わせる

**Reply by ta-ichihashi** [reply]
ありがとうございます。混流生産に対応できるようにする、という目標はその通りです。
ただ、本フレームワークではMESそのものの役割ではなくもう少し実行系に近い部分を目指しています。例に示していただいた品種Aを10, 品種Bを20という生産計画レベルは上位で処理いただければと思いますが、本フレームワークでは実際に材料（ワーク）がロードされた後、該当ジョブをどのように取り扱うか、からを問題領域にします。
とはいえ、材料をロードした後にバッファを持つ機構が有る場合や処理ステップの違い等で、追い越し、逆戻り、など最適スケジューリングについて考慮することは多々ありそうですね。その計画部分はいろいろなアルゴリズムのモジュールを付け替えられるように切り離し、材料ジョブからその計画に基づいて機械ジョブを発行する、という実行の仕組みのインターフェースをまずは作っていけたらと思います。

**Comment by ta-ichihashi** [comment]
よくよく考えると、材料ジョブに関しては、リアルタイムの制御側が責任を負うよりも、汎用OS側で計画を立てて、マシンジョブに変換して指示を出す方が効率的では？と思い始めました。あくまでも制御レイヤはマシンの構成に準じてジョブを抽象化することに専念し、汎用OS側（ADSやOPC UAを介したLinuxやWindows側）で、ワークジョブに準じてマシンジョブの実行手順を組み立ててスクリプトで指令する方が汎用性があるのではないかと。例えるならば、docker-composeでオーケストレーションを組み立てるようなイメージですかね。
スクリプトからマシンジョブの構成を生成するような機能を実現するには、factory methodデザインパターンなどを使うと良いのかもしれませんので、この方法で実現できないか検討を進めます。

**Comment by makitagawa243** [comment]
ジョブの単位間ではmilli/nanosecオーダーのリアルタイム性は必要無いと思いますので、
上位側で管理する方が柔軟性はありそうですね。
リアルタイム制御レイヤーの状態・機能を抽象化するところの実現方法で工夫が必要だと思います。
上位で凝ったことをやろうと思うと下位の情報を事細かに把握する必要が出てきて、どんどん密結合に寄っていってしまう経験は複数回あります。

**Comment by ta-ichihashi** [comment]
ジョブの新規生成、実行シナリオで一定のAPIは用意できたように思います。一度整理して今後の開発課題を整理しました。下図のとおり、機械ジョブは赤枠で、材料ジョブは青枠で書いてみました。

![image](https://github.com/user-attachments/assets/b0e97eb0-1185-4cc5-a4de-7373a61fa24a)

横軸は工程を表し、縦軸は時間を示します。各工程では時間軸に従って「機械ジョブ」（赤枠）が発行されます。これらのジョブは要求に応じて動的に生成されますのでQueueコンテナを用います、しかしこのコンテナに登録するジョブそのものは品種によりバリエーションがあるものの、再利用可能な動作パターンが繰り返されます。よって、 fa0eee0 で追加した `append_job` を使って事前に静的に生成したジョブを何度も再利用してQueueに登録して実行します。このモデルが適用できるのは、各工程が物理的にも同時実行できないことが条件となります。

これに対して「材料ジョブ」（青枠）は、やはり QUEUEコンテナで構成しますが、ここに構成するジョブの実体は、次のアクションを行う間接的な制御アクションです。

1. 材料ジョブのレシピに設定された品種に基づいて事前生成された静的な「機械ジョブ」を選択
2. 工程のQueueコンテナ（赤枠）に対してappend_jobする。同時にオブザーバパターンにより機械ジョブを監視開始する
3. 機械ジョブの終了を検出したらジョブ終了を通知する

次の開発目標は、この間接的な制御アクションを実行するジョブを内蔵しておく事です。

**Comment by makitagawa243** [comment]
平たくいうと、

* GEM300のProcess Jobライクなものが青枠
* 赤枠の各機械ジョブが処理ユニット個別のレシピ

ということですね。
材料ジョブを動的に組み立てられるため、材料投入順序にそった逐次実行だけでなく後続の追い越しなども実現できそうです。（大変ですが…）

**Reply by ta-ichihashi** [reply]
その通りです。どの材料ジョブをどのタイミングで実行するかはまた別問題ですかね。スケジューリングアルゴリズムは極めて難しい課題です。ただ、オブザーバパターンで各ジョブの実行ログは取れますので、実績時間やそのジッタなどの傾向からある程度予測は正確にできるかもしれません。

**Comment by masudako** [comment]
GEM300は、半導体業界オリジナルの通信プロトコルなのでしょうか・・・。知識が乏しくて申し訳ありません。
理解として、下記のようなイメージでしょうか。
　■機械ジョブ（赤枠のコンテナ）に関しては、各設備モジュール（圧入モジュールとか洗浄モジュールとか）が
　　実際に実行するジョブを管理する。（各設備モジュールは、機械ジョブに従って順次動作する）
　■一方、材料ジョブ（青枠のコンテナ）は、実際の物を動かす役目では無く、機械ジョブの構築を役割として担う。
　　機械ジョブの完了を持って材料ジョブは副次的に進行していく事で、材料のプロセス進行を表示する事ができる。

ちなみに、材料ジョブが扱うワークの単位としては、ワーク一つ毎の単位でしょうか？
それとも、ロット単位の、同一レシピで生産できる固まりを示すイメージでしょうか？

**Reply by makitagawa243** [reply]
> GEM300は、半導体業界オリジナルの通信プロトコル

ご認識の通りです。非常に膨大な規格でして私も奥深くまでは把握できておりません…

機械ジョブ・材料ジョブについても、同じ理解をしております。
材料ジョブを再帰構造にすることで、ワークを内包したロット単位も表現可能と考えます。

**Comment by masudako** [comment]
> 材料ジョブを再帰構造にすることで、ワークを内包したロット単位も表現可能と考えます。

あ、そゆことですか。なんとなく理解が出来た気がします。ありがとうございます。

**Comment by ta-ichihashi** [comment]
もう少し抽象化してモデル分析すると、主従モデルですかね。材料加工ジョブは主ジョブ、機械ジョブは従ジョブと読み替えられます。そして主ジョブが従ジョブに実行を移譲する、という解釈もできます。この主従モデルは、機械とワークとの間だけでなく、XTSやXPlanarのようなトラック搬送システムに対しても適用できますね。搬送元から搬送先への経路（＝トラック）搬送が主ジョブで、個々の可動子に与える動作指令が従ジョブとなります。さらに材料ジョブは主ジョブとなって、経路搬送のジョブに指示を与える、といった関係が構築できます。ワーク、トラック、搬送装置、それぞれの責務が違うので、疎結合しておくことで組合せ自由度が高まると思います。
上司部下で役割分担して効率的にリソースを動かす感じですかね。ソフトウェアはやはり人間社会モデルと同じなのかもしれません。
